doc/developer/design/20230714_show_privileges.md

```diff
@@ -0,0 +1,266 @@
+- [RBAC design doc](20230216_role_based_access_control.md)
+- [System privilege design doc](20230531_system_privileges.md)
+- [Default privilege design doc](20230602_default_privileges.md)
+
+## Context
+
+Role based access control comprises a set of features that allow customers to control the
+privileges of individual users. Privileges can be granted to a user which allows them to take
+certain actions in a Materialize deployment. All objects have an owner, and certain actions are
+limited to only object owners. Roles can be members of other roles and inherit all of their
+privileges. Privileges, owners, and role membership are all stored within the database catalog.
+
+Currently, it's difficult for a user to discover RBAC related information.
+
+## Goals
+
+- Make it easy for users to discover the set of privileges that they have.
+- Make it easy for users to discover the owner of objects.
+- Make it easy for users to discover the roles that they are a member of.
+
+## Overview
+
+We will add the following SQL commands:
+
+- `SHOW PRIVILEGES`: shows the privileges of the current user.
+- `SHOW DEFAULT PRIVILEGES`: shows default privileges.
+- `SHOW ROLE MEMBERSHIPS`
+
+We will also add an `owner` column to the existing show commands.
+
+## Detailed description
+
+### `SHOW PRIVILEGES`
+
+Will show object privileges.
+
+The syntax will be: `SHOW [<object-type>] PRIVILEGES [FOR {<role> | ALL ROLES}] [FROM {SCHEMA <schema-name> | ALL SCHEMAS | DATABASE <database> | ALL DATABASES}]`.
+
+- `<object-type>` will filter the output to only include objects of type `<object-type>`. Valid
+values are:
+  - `OBJECT`
+  - `DATABASE`
+  - `SCHEMA`
+  - `CLUSTER`
+  - `SYSTEM`
+- `FOR <role>` will filter the output to only include privileges granted to `<role>`.
+  - If a role or `ALL ROLES` is not specified then the current role is assumed.
+- `FOR ALL ROLES` will display privileges granted to all roles.
+- `FROM SCHEMA <schema-name>` will filter the output to exclude items not in schema `<schema-name>`.
+  - This has no effect on database, schema, cluster, or system privileges.
+  - If a schema or `ALL SCHEMAS` is not specified then the active schema is assumed.
+  - Invalid if an object type of `DATABASE`, `SCHEMA`, `CLUSTER`, or `SYSTEM` has been specified.
+- `FROM ALL SCHEMAS` will include all items.
+  - Invalid if an object type of `DATABASE`, `SCHEMA`, `CLUSTER`, or `SYSTEM` has been specified.
+- `FROM DATABASE <database>` will filter the output to exclude schemas not in database `<database>`.
+  - This has no effect on item, database, cluster, or cluster privileges.
+  - If a database or `ALL DATABASES` is not specified then the active database is assumed.
+  - Invalid if an object type of `DATABASE`, `CLUSTER`, or `SYSTEM` has been specified. 
+- `FROM ALL DATABASES` will include all schemas.
+  - Invalid if an object type of `DATABASE`, `CLUSTER`, or `SYSTEM` has been specified. 
+
+The output will include the following columns:
```

Comment:
Do we need a `role` column now? `SHOW PRIVILEGES FOR ALL ROLES` needs a way to identify which privileges are for which role?

In that case, I wonder what we put in that column for `SHOW PRIVILEGES` - is it the user's role or is the parent role from which they inherited the privilege from? Thinking the user's role is what the user would expect.

Then in `SHOW PRIVILEGES FOR ALL ROLES` I guess we show the total privileges of each role, both directly granted and inherited? Is there a query the user could run to figure out which privileges were directly granted vs inherited?

Reply:
> Do we need a `role` column now? `SHOW PRIVILEGES FOR ALL ROLES` needs a way to identify which privileges are for which role?

Yes good point, I've added both a `grantee` and a `grantor` column.

> In that case, I wonder what we put in that column for `SHOW PRIVILEGES` - is it the user's role or is the parent role from which they inherited the privilege from? Thinking the user's role is what the user would expect.

I think that it's actually best to use the parent role from which they inherited the privilege from. If it's always the user role, then each row will have the same value for that column and it wouldn't be adding any additional useful information. For example if you had

```
SHOW PRIVILEGES FOR r1;

 database    | schema | name        | object_type | privileges     | role
-------------+--------+-------------+-------------+---------+----------------+---------
 materialize | public | my_table    | table       | SELECT, INSERT | r1
 materialize | public | my_view     | view        | SELECT         | r1
 materialize | NULL   | public      | schema      | USAGE          | r1
 NULL        | NULL   | materialize | database    | USAGE          | r1
 NULL        | NULL   | my_db       | database    | USAGE          | r1
 NULL        | NULL   | my_cluster  | cluster     | CREATE         | r1
```
The `role` column is not very useful and you have no way of knowing where the grant actually comes from. The parent role also matches with the existing PostgreSQL information schema tables on privileges.

> Then in `SHOW PRIVILEGES FOR ALL ROLES` I guess we show the total privileges of each role, both directly granted and inherited? Is there a query the user could run to figure out which privileges were directly granted vs inherited?

I think this becomes too noisy and we should only show privileges granted directly. Otherwise each privilege gets multiplied by the amount of times it's inherited. If the user want's to know a particular roles effective privilege, then they could run `SHOW PRIVILEGES FOR ROLE <role>`.